wolfSSL with Doxygen Documentation

connerWolfSSL opened the pull request:
supersedes PR #1348 
changed project branch
Made recommended changes from previous PR

dgarske commented on doc/README_DOXYGEN:
```diff
@@ -0,0 +1,182 @@
+wolfSSL with Doxygen 1.8.13
```
Please run a spell check on this. `comands`, `secions`, `seperated`, `especialy`, `eachother`.

connerWolfSSL commented:
Thanks again for the feedback. I'll be making changes today and hopefully
pushing either today or on Tuesday. Yes, all API in
dox_comment/header_files was pulled out the manual by hand. There were to
many discrepancies to create a script to sift through the manual. In
regards to doxygen warning if API  that was previously documented is
changed it currently does not. As far as doxygen is concerned the API in
dox_comment/header_files is unique to the API in the main wolfssl
directory. I am working on a script that will check if all API prototypes
in the dox_headers/header_files directory are present within the main
wolfssl directory as well. As for adding new API to the library, any new
API with documentation will have to be manually added to the header_files
directory. The header_files directory only contains API that currently has
documentation. Undocumented API does not belong in the
dox_comments/header_files directory.

I piped all errors/warnings to separate files to help navigate the large
walls of text that are output by doxygen and latex. The scripts doxygen and
latex use are very verbose. I have attempted cleaning up many of the
warnings output (especially when generating the pdf), but I believe they
mostly stem from items generated by doxygen. If you would prefer everything
output to the console I can change it back.

I will double check all the latex dependencies again. It's possible I
missed a few due to packages already being present on my machine before
this project, therefor I didn't receive the same errors as you.

On Tue, Feb 20, 2018 at 4:53 PM, John Safranek <notifications@github.com>
wrote:

> *@ejohnstown* commented on this pull request.
>
> As a followup for David's comment on which LaTeX install to add, for the
> Mac I'm recommending Basic TeX from TUG. It is considerably smaller than
> MacTeX or TeX Live. It has an installer, but it is missing some of the TeX
> packages. Those can be installed with:
> % sudo tlmgr update --self
> % sudo tlmgr install tabu varwidth multirow adjustbox collectbox sectsty
> tocloft collection-fontsextra
> After all that installed, I was able to build the PDF of the manual.
>
> I didn't need LaTeX for the HTML version.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wolfSSL/wolfssl/pull/1361#pullrequestreview-98034300>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbYcWUQ1y2PR_iAkahL_q7KRGKGP-lz1ks5tW1r1gaJpZM4R-ubZ>
> .
>

connerWolfSSL commented:
I have added a script to check the documented API against the wolfSSL library. I have tested it and believe it is working (it already caught a few discrepancies as well). Feel free to test it yourself as well. The scripts name is check_api.sh. I also removed the piping of errors and warnings to files and instead let everything output to the terminal.

dgarske commented on doc/generate_documentation.sh:
```diff
@@ -0,0 +1,73 @@
+exit 1
+fi
+
+if [ ! -e "build/bin/doxygen" ]; then
```
Could we change this to just be a separate option like `./generate_documentation.sh -install`?

I think you could also detect that `doxygen` is missing on the system using something like `which doxygen`. If its missing then fire off the git clone and build and set local PATH variable for `./build/bin`.

When it tries to build doxygen on my Mac I get errors with the `ICONV_ACCEPTS_CONST_INPUT` test.

On the Mac I can easily install it using `brew install doxygen` and it puts it into `/usr/local/Cellar/doxygen/1.8.14/bin/`, which I can add to my path using `export PATH="/usr/local/Cellar/doxygen/1.8.14/bin/:$PATH"`.

Then below call using `doxygen` instead of `build/bin/doxygen`.
